Mathlib/Data/Zmod/Defs.lean

```diff
@@ -0,0 +1,188 @@
+/-
+Copyright (c) 2022 Eric Rodriguez. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Eric Rodriguez
+
+! This file was ported from Lean 3 source module data.zmod.defs
+! leanprover-community/mathlib commit 1126441d6bccf98c81214a0780c73d499f6721fe
+! Please do not edit these lines, except to modify the commit id
+! if you have ported upstream changes.
+-/
+import Mathlib.Algebra.NeZero
+import Mathlib.Data.Nat.ModEq
+import Mathlib.Data.Fintype.Lattice
+
+/-!
+# Definition of `Zmod n` + basic results.
+
+This file provides the basic details of `Zmod n`, including its commutative ring structure.
+
+## Implementation details
+
+This used to be inlined into `Data.Zmod.Basic`. This file imports `CharP.Basic`, which is an
+issue; all `CharP` instances create an `Algebra (Zmod p) R` instance; however, this instance may
+not be definitionally equal to other `Algebra` instances (for example, `GaloisField` also has an
+`Algebra` instance as it is defined as a `SplittingField`). The way to fix this is to use the
+forgetful inheritance pattern, and make `CharP` carry the data of what the `smul` should be (so
+for example, the `smul` on the `GaloisField` `CharP` instance should be equal to the `smul` from
+its `SplittingField` structure); there is only one possible `Zmod p` algebra for any `p`, so this
+is not an issue mathematically. For this to be possible, however, we need `CharP.Basic` to be
+able to import some part of `Zmod`.
+
+-/
+
+
+namespace Fin
+
+/-!
+## Ring structure on `Fin n`
+
+We define a commutative ring structure on `Fin n`, but we do not register it as instance.
+Afterwords, when we define `Zmod n` in terms of `Fin n`, we use these definitions
+to register the ring structure on `Zmod n` as type class instance.
```

Comment:
Feel free to rewrite, and maybe leave a TODO that we should streamline this file, now that `Fin` has more instances.

Reply:
I just dropped the obviously false bit